File: src/PythonClient/lib/quixstreaming/models/parameterdatatimestamp.py
from typing import Union, Dict

from quixstreaming.helpers import *
from quixstreaming.models.netdict import ReadOnlyNetDict
from datetime import datetime, timedelta

from quixstreaming.models.parametervalue import ParameterValue, ParameterValueType
from .. import __importnet
import Quix.Sdk.Streaming
import clr
clr.AddReference('System.Collections')
import System.Collections.Generic


class ParameterDataTimestamp(object):
    """
    Represents a single point in time with parameter values and tags attached to that time
    """

    def __init__(self, net_object: Quix.Sdk.Streaming.Models.ParameterDataTimestamp):
        """
            Initializes a new instance of ParameterDataTimestamp.

            Parameters:

            net_object (.net object): The .net object representing ParameterDataTimestamp.
        """
        self.__wrapped = net_object

    def __str__(self):
        text = "Time:" + str(self.timestamp_nanoseconds)
        text += "\r\n  Tags: " + str(self.tags)
        text += "\r\n  Params:"
        for param_id, param_val in self.parameters.items():
            if param_val.type == ParameterValueType.Numeric:
                text += "\r\n    " + str(param_id) + ": " + str(param_val.numeric_value)
                continue
            if param_val.type == ParameterValueType.String:
                text += "\r\n    " + str(param_id) + ": " + str(param_val.string_value)
                continue
            if param_val.type == ParameterValueType.Binary:
                text += "\r\n    " + str(param_id) + ": byte[" + str(len(param_val.binary_value)) + "]"
                continue
            if param_val.type == ParameterValueType.Empty:
                text += "\r\n    " + str(param_id) + ": Empty"
                continue
            text += "\r\n    " + str(param_id) + ": ???"
        return text

    @property
    def parameters(self) -> Dict[str, ParameterValue]:
        """
        Parameter values for the timestamp. When a key is not found, returns empty ParameterValue
        The dictionary key is the parameter id
        The dictionary value is the value (ParameterValue)
        """

        def _value_converter_to_python(net_object: Quix.Sdk.Streaming.Models.ParameterValue):
            if net_object is None:
                return None
            return ParameterValue(net_object)

        def _value_converter_from_python(param_val: ParameterValue):
            if param_val is None:
                return None
            return param_val.convert_to_net()

        return ReadOnlyNetDict(self.__wrapped.Parameters, val_converter_from_python=_value_converter_from_python,
                               val_converter_to_python=_value_converter_to_python)

    @property
    def tags(self) -> Dict[str, str]:
        """
        Tags for the timestamp. When key is not found, returns None
        The dictionary key is the tag id
        The dictionary value is the tag value
        """

        return ReadOnlyNetDict(self.__wrapped.Tags)

    @property
    def timestamp_nanoseconds(self) -> int:
        """Gets timestamp in nanoseconds"""

        return self.__wrapped.TimestampNanoseconds

    @property
    def timestamp_milliseconds(self) -> int:
        """Gets timestamp in milliseconds"""

        return self.__wrapped.TimestampMilliseconds

    @property
    def timestamp(self) -> datetime:
        """Gets the timestamp in datetime format"""
        return NetToPythonConverter.convert_datetime(self.__wrapped.Timestamp)

    @property
    def timestamp_as_time_span(self) -> timedelta:
        """Gets the timestamp in timespan format"""
        return NetToPythonConverter.convert_timespan(self.__wrapped.TimestampAsTimeSpan)

    def add_value(self, parameter_id: str, value: Union[float, str, int, bytearray, bytes]) -> 'ParameterDataTimestamp':
        """
            Adds a new value for the parameter
            :param parameter_id: The parameter to add the value for
            :param value: the value to add. Can be float or string

        :return: ParameterDataTimestamp
        """

        if type(value) is int:
            value = float(value)

        self.__wrapped.AddValue(parameter_id, value)
        return self

    def remove_value(self, parameter_id: str)  -> 'ParameterDataTimestamp':
        """
            Removes the value for the parameter
            :param parameter_id: The parameter to remove the value for

        :return: ParameterDataTimestamp
        """
        self.__wrapped.RemoveValue(parameter_id)
        return self

    def add_tag(self, tag_id: str, tag_value: str)  -> 'ParameterDataTimestamp':
        """
            Adds a tag to the values
            :param tag_id: The id of the tag to the set the value for
            :param tag_value: the value to set

        :return: ParameterDataTimestamp
        """
        self.__wrapped.AddTag(tag_id, tag_value)
        return self

    def remove_tag(self, tag_id: str)  -> 'ParameterDataTimestamp':
        """
            Removes a tag from the values
            :param tag_id: The id of the tag to remove

        :return: ParameterDataTimestamp
        """
        self.__wrapped.RemoveTag(tag_id)
        return self

    def add_tags(self, tags: Dict[str, str]) -> 'ParameterDataTimestamp':
        """
            Copies the tags from the specified dictionary. Conflicting tags will be overwritten
            :param tags: The tags to add

        :return: ParameterDataTimestamp
        """

        if tags is None:
            return self

        prep_tags_dict = System.Collections.Generic.List[System.Collections.Generic.KeyValuePair[str, str]]([])
        for key, val in tags.items():
            prep_tags_dict.Add(System.Collections.Generic.KeyValuePair[str, str](key, val))
        self.__wrapped.AddTags(prep_tags_dict)
        return self

    def convert_to_net(self):
        return self.__wrapped

File: src/CsharpClient/Quix.Sdk.Streaming/Models/ParameterDataTimestamp.cs
using System;
using System.Collections.Generic;
using Quix.Sdk.Process.Models.Utility;
using Quix.Sdk.Streaming.Utils;

namespace Quix.Sdk.Streaming.Models
{
    /// <summary>
    /// Represents a single point in time with parameter values and tags attached to that time
    /// </summary>
    public readonly struct ParameterDataTimestamp
    {
        internal readonly ParameterData parameterData;
        internal readonly long timestampRawIndex;

        internal ParameterDataTimestamp(ParameterData parameterData, long rawIndex)
        {
            this.parameterData = parameterData;
            this.timestampRawIndex = rawIndex;
        }

        /// <summary>
        /// Parameter values for the timestamp. When a key is not found, returns empty <see cref="ParameterValue"/>
        /// </summary>
        public readonly ParameterDataTimestampValues Parameters => new ParameterDataTimestampValues(this.parameterData, this.timestampRawIndex);

        /// <summary>
        /// Tags for the timestamp. When key is not found, returns null
        /// </summary>
        public readonly ParameterDataTimestampTags Tags  => new ParameterDataTimestampTags(this.parameterData, this.timestampRawIndex);

        /// <summary>
        /// Gets the timestamp in nanoseconds
        /// </summary>
        public long TimestampNanoseconds
        {
            get
            {
                return !this.parameterData.epochsIncluded[this.timestampRawIndex] 
                    ? this.parameterData.rawData.Timestamps[this.timestampRawIndex] + this.parameterData.rawData.Epoch
                    : this.parameterData.rawData.Timestamps[this.timestampRawIndex];
            }
            set
            {
                this.parameterData.rawData.Timestamps[this.timestampRawIndex] = value;
            }
        }

        /// <summary>
        /// Gets the timestamp in milliseconds
        /// </summary>
        public long TimestampMilliseconds => this.TimestampNanoseconds / (long) 1e6;

        /// <summary>
        /// Gets the timestamp in <see cref="DateTime"/> format
        /// </summary>
        public DateTime Timestamp => this.TimestampNanoseconds.FromUnixNanoseconds();

        /// <summary>
        /// Gets the timestamp in <see cref="TimeSpan"/> format
        /// </summary>
        public TimeSpan TimestampAsTimeSpan => this.TimestampNanoseconds.FromNanoseconds();

        internal bool EpochIncluded
        {
            get
            {
                return this.parameterData.epochsIncluded[this.timestampRawIndex];
            }
            set
            {
                this.parameterData.epochsIncluded[this.timestampRawIndex] = value;
            }
        }

        /// <summary>
        /// Adds a new numeric value.
        /// </summary>
        /// <param name="parameterId">Parameter Id</param>
        /// <param name="value">Numeric value</param>
        /// <returns>This instance</returns>
        public ParameterDataTimestamp AddValue(string parameterId, double value)
        {
            if (!this.parameterData.rawData.NumericValues.TryGetValue(parameterId, out var values))
            {
                values = new double?[this.parameterData.rawData.Timestamps.Length];
                this.parameterData.rawData.NumericValues.Add(parameterId, values);

                this.parameterData.parameterList[parameterId] = new Parameter(parameterId, values);
            }

            values[this.timestampRawIndex] = value;

            return this;
        }

        /// <summary>
        /// Adds a new string value.
        /// </summary>
        /// <param name="parameterId">Parameter Id</param>
        /// <param name="value">String value</param>
        /// <returns>This instance</returns>
        public ParameterDataTimestamp AddValue(string parameterId, string value)
        {
            if (!this.parameterData.rawData.StringValues.TryGetValue(parameterId, out var values))
            {
                values = new string[this.parameterData.rawData.Timestamps.Length];
                this.parameterData.rawData.StringValues.Add(parameterId, values);

                this.parameterData.parameterList[parameterId] = new Parameter(parameterId, values);
            }

            values[this.timestampRawIndex] = value;

            return this;
        }

        /// <summary>
        /// Adds a new binary value.
        /// </summary>
        /// <param name="parameterId">Parameter Id</param>
        /// <param name="value">String value</param>
        /// <returns>This instance</returns>
        public ParameterDataTimestamp AddValue(string parameterId, byte[] value)
        {
            if (!this.parameterData.rawData.BinaryValues.TryGetValue(parameterId, out var values))
            {
                values = new byte[this.parameterData.rawData.Timestamps.Length][];
                this.parameterData.rawData.BinaryValues.Add(parameterId, values);

                this.parameterData.parameterList[parameterId] = new Parameter(parameterId, values);
            }

            values[this.timestampRawIndex] = value;

            return this;
        }

        /// <summary>
        /// Adds a new value.
        /// </summary>
        /// <param name="parameterId">Parameter Id</param>
        /// <param name="value">The value</param>
        /// <returns>This instance</returns>
        public ParameterDataTimestamp AddValue(string parameterId, ParameterValue value)
        {
            if (value.Value == null) return this;

            var valueType = value.Type;

            if (valueType == ParameterValueType.Numeric) this.AddValue(parameterId, value.NumericValue ?? 0);
            else if (valueType == ParameterValueType.String) this.AddValue(parameterId, value.StringValue);
            else if (valueType == ParameterValueType.Binary) this.AddValue(parameterId, value.BinaryValue);

            return this;
        }

        /// <summary>
        /// Removes a parameter value.
        /// </summary>
        /// <param name="parameterId">Parameter Id</param>
        /// <returns>This instance</returns>
        public ParameterDataTimestamp RemoveValue(string parameterId)
        {
            var parameter = this.Parameters[parameterId];
            var valueType = parameter.Type;

            if (valueType == ParameterValueType.Numeric) parameter.NumericValue = null;
            else if (valueType == ParameterValueType.String) parameter.StringValue = null;
            else if (valueType == ParameterValueType.Binary) parameter.BinaryValue = null;

            return this;
        }

        /// <summary>
        /// Adds a tag to the values
        /// </summary>
        /// <param name="tagId">Tag name</param>
        /// <param name="tagValue">Tag value</param>
        /// <returns>This instance</returns>
        public ParameterDataTimestamp AddTag(string tagId, string tagValue)
        {
            if (string.IsNullOrWhiteSpace(tagId)) throw new ArgumentNullException(nameof(tagId), "Tag id can't be null or empty");
            if (string.IsNullOrWhiteSpace(tagValue)) throw new ArgumentNullException(nameof(tagValue), $"Tag ({tagId}) value can't be null or empty");

            if (!this.parameterData.rawData.TagValues.TryGetValue(tagId, out var values))
            {
                values = new string[this.parameterData.rawData.Timestamps.Length];
                this.parameterData.rawData.TagValues.Add(tagId, values);
            }

            values[this.timestampRawIndex] = tagValue;

            return this;
        }
        
        /// <summary>
        /// Copies the tags from the specified dictionary.
        /// Conflicting tags will be overwritten
        /// </summary>
        /// <param name="tags">The tags to copy</param>
        /// <returns>This instance</returns>
        public ParameterDataTimestamp AddTags(IEnumerable<KeyValuePair<string, string>> tags)
        {
            if (tags == null) return this;

            foreach (var tagPair in tags)
            {
                this.AddTag(tagPair.Key, tagPair.Value);
            }

            return this;
        }
        /// <summary>
        /// Removes a tag from the values
        /// </summary>
        /// <param name="tagId">Tag name</param>
        /// <returns>This instance</returns>
        public ParameterDataTimestamp RemoveTag(string tagId)
        {
            this.parameterData.rawData.TagValues.Remove(tagId);

            return this;
        }
    }
}